File: test/src/tree/attribute_tree/newick_reader.cpp
/*
    Genesis - A toolkit for working with phylogenetic data.
    Copyright (C) 2014-2017 Lucas Czech

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    Contact:
    Lucas Czech <lucas.czech@h-its.org>
    Exelixis Lab, Heidelberg Institute for Theoretical Studies
    Schloss-Wolfsbrunnenweg 35, D-69118 Heidelberg, Germany
*/

/**
 * @brief Testing Default Tree methods.
 *
 * @file
 * @ingroup test
 */

#include "src/common.hpp"

#include "genesis/tree/attribute_tree/tree.hpp"
#include "genesis/tree/attribute_tree/indexed_newick_reader.hpp"
#include "genesis/tree/attribute_tree/keyed_newick_reader.hpp"
#include "genesis/tree/function/functions.hpp"
#include "genesis/tree/formats/newick/reader.hpp"
#include "genesis/tree/tree.hpp"

#include <algorithm>
#include <string>
#include <utility>

using namespace genesis;
using namespace tree;

std::pair<size_t, size_t> count_attribute_tree_data( AttributeTree const& tree )
{
    size_t node_attr_cnt = 0;
    for( auto const& node : tree.nodes() ) {
        auto const& data = node->data<AttributeTreeNodeData>();
        node_attr_cnt += data.attributes.size();
    }

    size_t edge_attr_cnt = 0;
    for( auto const& edge : tree.edges() ) {
        auto const& data = edge->data<AttributeTreeEdgeData>();
        edge_attr_cnt += data.attributes.size();
    }

    return { node_attr_cnt, edge_attr_cnt };
}

void print_attribute_tree_data( AttributeTree const& tree )
{
    LOG_DBG << "Nodes";
    for( auto const& node : tree.nodes() ) {
        auto const& data = node->data<AttributeTreeNodeData>();
        LOG_DBG1 << "node " << data.name;
        for( auto const& m : data.attributes ) {
            LOG_DBG2 << m.first << " --> " << m.second;
        }
    }

    LOG_DBG << "Edges";
    for( auto const& edge : tree.edges() ) {
        auto const& data = edge->data<AttributeTreeEdgeData>();
        LOG_DBG1 << "edge";
        for( auto const& m : data.attributes ) {
            LOG_DBG2 << m.first << " --> " << m.second;
        }
    }
}

TEST( AttributeTree, IndexedNewickReaderIndex )
{
    // Skip test if no data availabe.
    NEEDS_TEST_DATA;

    // Read and process tree.
    std::string infile = environment->data_dir + "tree/indexed_attributes_0.newick";

    IndexedAttributeTreeNewickReader reader;
    reader.add_attribute(
        IndexedAttributeTreeNewickReader::Source::kComment, 0,
        IndexedAttributeTreeNewickReader::Target::kEdge,    "bootstrap"
    );

    auto tree = reader.from_file( infile );

    auto counts = count_attribute_tree_data( tree );
    EXPECT_EQ( 0, counts.first );
    EXPECT_EQ( 3, counts.second );

    // print_attribute_tree_data( tree );
}

TEST( AttributeTree, IndexedNewickReaderCatchAll )
{
    // Skip test if no data availabe.
    NEEDS_TEST_DATA;

    // Read and process tree.
    std::string infile = environment->data_dir + "tree/indexed_attributes_1.newick";

    IndexedAttributeTreeNewickReader reader;
    reader.add_catch_all(
        IndexedAttributeTreeNewickReader::Source::kComment,
        IndexedAttributeTreeNewickReader::Target::kEdge,    "comment_"
    );

    auto tree = reader.from_file( infile );

    size_t node_attr_cnt = 0;
    for( auto const& node : tree.nodes() ) {
        auto const& data = node->data<AttributeTreeNodeData>();
        node_attr_cnt += data.attributes.size();
    }

    size_t edge_attr_cnt = 0;
    for( auto const& edge : tree.edges() ) {
        auto const& data = edge->data<AttributeTreeEdgeData>();
        edge_attr_cnt += data.attributes.size();
    }

    auto counts = count_attribute_tree_data( tree );
    EXPECT_EQ(  0, counts.first );
    EXPECT_EQ( 12, counts.second );
}

TEST( AttributeTree, KeyedNewickReaderKeys )
{
    // Skip test if no data availabe.
    NEEDS_TEST_DATA;

    // Read and process tree.
    std::string infile = environment->data_dir + "tree/keyed_attributes_0.newick";

    KeyedAttributeTreeNewickReader reader;
    reader.add_attribute( "bs",     KeyedAttributeTreeNewickReader::Target::kEdge );
    reader.add_attribute( "!color", KeyedAttributeTreeNewickReader::Target::kEdge, "color" );

    auto tree = reader.from_file( infile );

    auto counts = count_attribute_tree_data( tree );
    EXPECT_EQ( 0, counts.first );
    EXPECT_EQ( 4, counts.second );
}

TEST( AttributeTree, KeyedNewickReaderCatchAll )
{
    // Skip test if no data availabe.
    NEEDS_TEST_DATA;

    // Read and process tree.
    std::string infile = environment->data_dir + "tree/keyed_attributes_0.newick";

    KeyedAttributeTreeNewickReader reader;
    reader.add_catch_all( KeyedAttributeTreeNewickReader::Target::kEdge );

    auto tree = reader.from_file( infile );

    auto counts = count_attribute_tree_data( tree );
    EXPECT_EQ( 0, counts.first );
    EXPECT_EQ( 4, counts.second );

    // print_attribute_tree_data( tree );
}

TEST( AttributeTree, KeyedNewickReaderNHX )
{
    // Skip test if no data availabe.
    NEEDS_TEST_DATA;

    // Read and process tree.
    std::string infile = environment->data_dir + "tree/keyed_attributes_1.newick";

    KeyedAttributeTreeNewickReader reader;
    reader.add_nhx_attributes();
    // reader.add_catch_all( KeyedAttributeTreeNewickReader::Target::kNode );

    auto tree = reader.from_file( infile );

    auto counts = count_attribute_tree_data( tree );
    EXPECT_EQ( 25, counts.first );
    EXPECT_EQ(  1, counts.second );

    // print_attribute_tree_data( tree );
}

File: python/giannis/test/src/tree/attribute_tree/newick_reader.py
import sys
import unittest

sys.path.append('./genesis_bindings')

from mylibgenesis.genesis.utils import *
from mylibgenesis.genesis.tree import *

class NewickReaderTest(unittest.TestCase):

	def count_attribute_tree_data( self, tree: Tree):
		node_attr_cnt = 0
		for node in tree.nodes():
			data = node.data()
			node_attr_cnt = node_attr_cnt + len(data.attributes)

		edge_attr_cnt = 0
		for edge in tree.edges():
			data = edge.data()
			edge_attr_cnt = edge_attr_cnt + len(data.attributes)

		return (node_attr_cnt, edge_attr_cnt)

	def print_attribute_tree_data( self, tree: Tree):
		#LOG_DBG << "Nodes";
		for node in tree.nodes():
			data = node.data()
			#LOG_DBG1 << "node " << data.name;
			for m in data.attributes:
				True
				#LOG_DBG2 << m.first << " --> " << m.second;

		#LOG_DBG << "Edges";
		for edge in tree.edges():
			data = edge.data();
			#LOG_DBG1 << "edge";
			for m in data.attributes:
				True
				#LOG_DBG2 << m.first << " --> " << m.second;

	def test_indexed_newick_reader_index(self):

		#Read and process tree.
		infile = "test/data/tree/indexed_attributes_0.newick"

		reader = IndexedAttributeTreeNewickReader()
		reader.add_attribute( IndexedAttributeTreeNewickReader.Source.kComment, 0, IndexedAttributeTreeNewickReader.Target.kEdge, "bootstrap")

		tree = reader.read( from_file( infile ))

		counts = self.count_attribute_tree_data( tree )
		self.assertEqual(0,counts[0])
		self.assertEqual(3,counts[1])

		#print_attribute_tree_data( tree );

	def test_indexed_newick_reader_catch_all(self):

		# Read and process tree.
		infile = "test/data/tree/indexed_attributes_1.newick"

		reader = IndexedAttributeTreeNewickReader()
		reader.add_catch_all( IndexedAttributeTreeNewickReader.Source.kComment, IndexedAttributeTreeNewickReader.Target.kEdge, "comment_")

		tree = reader.read( from_file( infile ))

		node_attr_cnt = 0
		for node in tree.nodes():
			data = node.data();
			node_attr_cnt = node_attr_cnt + len(data.attributes)
		self.assertEqual(0,node_attr_cnt)

		edge_attr_cnt = 0
		for edge in tree.edges():
			data = edge.data()
			edge_attr_cnt += len(data.attributes)
		self.assertEqual(12,edge_attr_cnt)

		counts = self.count_attribute_tree_data( tree )
		self.assertEqual(0,counts[0])
		self.assertEqual(12,counts[1])

	def test_keyed_newick_reader_keys(self):

		#Read and process tree.
		infile = "test/data/tree/keyed_attributes_0.newick"

		reader = KeyedAttributeTreeNewickReader()
		reader.add_attribute( "bs",     KeyedAttributeTreeNewickReader.Target.kEdge )
		reader.add_attribute( "!color", KeyedAttributeTreeNewickReader.Target.kEdge, "color" )

		tree = reader.read( from_file( infile ))

		counts = self.count_attribute_tree_data( tree )
		self.assertEqual(0,counts[0])
		self.assertEqual(4,counts[1])

	def test_keyed_newick_reader_catch_all(self):

		# Read and process tree.
		infile = "test/data/tree/keyed_attributes_0.newick"

		reader = KeyedAttributeTreeNewickReader()
		reader.add_catch_all( KeyedAttributeTreeNewickReader.Target.kEdge )

		tree = reader.read( from_file( infile ))

		counts = self.count_attribute_tree_data( tree )
		self.assertEqual(0,counts[0])
		self.assertEqual(4,counts[1])

		# print_attribute_tree_data( tree );

	def test_keyed_newick_reader_nhx(self):

		# Read and process tree.
		infile = "test/data/tree/keyed_attributes_1.newick"

		reader = KeyedAttributeTreeNewickReader()
		reader.add_nhx_attributes()
		# reader.add_catch_all( KeyedAttributeTreeNewickReader::Target::kNode );

		tree = reader.read( from_file( infile ))

		counts = self.count_attribute_tree_data( tree )
		self.assertEqual(25,counts[0])
		self.assertEqual(1,counts[1])

		# print_attribute_tree_data( tree );

if __name__ == '__main__':
	unittest.main()
